### PR TITLE
openstack: Use common SafeFormatAndMount() for mounting.

### LINKED ISSUE #17565: Cinder Volume Plugin Should use mount.SafeFormatAndMount
The Cinder volume plugin currently implements its own version of SafeFormatAndMount. It should use mount.SafeFormatAndMount from "util/mount"

@jsafrane @spothanis

### PR DESCRIPTION
**Pull request by jsafrane**
There is no reason for OpenStack + Cinder to use it's own variant of format
& mount.

Fixes #17565.

### CODE REVIEW

File: pkg/volume/cinder/cinder.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func (plugin *cinderPlugin) newBuilderInternal(spec *volume.Spec, podUID types.U
 		},
 		fsType:             fsType,
 		readOnly:           readOnly,
-		blockDeviceMounter: &cinderSafeFormatAndMount{mounter, exec.New()}}, nil
+		blockDeviceMounter: &mount.SafeFormatAndMount{plugin.host.GetMounter(), exec.New()}}, nil
```
Comment (swagiaal) [inline pkg/volume/cinder/cinder.go]: Hmm.. s/plugin.host.GetMounter()/mounter/
Not sure if the cinder tests use this but other volume plugin tests use this to pass in mock arguments

**Comment by swagiaal** [comment]
@k8s-bot unit test this

File: pkg/volume/cinder/cinder.go (hunk 1 above)
Comment (jsafrane) [inline pkg/volume/cinder/cinder.go]: Fixed. (Still, AWS has the same `plugin.host.GetMounter()` call)

Comment (swagiaal) [inline pkg/volume/cinder/cinder.go]: > Fixed. 

Great.. thanks

>(Still, AWS has the same plugin.host.GetMounter() call)

Ah! True...  That is because AWS and Cinder tests do not have this check:
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gce_pd/gce_pd_test.go#L169 :(

**Comment by swagiaal** [comment]
@jsafrane could you please rebase on master

**Comment by jsafrane** [comment]
Rebased

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -606,8 +606,8 @@ func validateCinderVolumeSource(cd *api.CinderVolumeSource) validation.ErrorList
 	if cd.VolumeID == "" {
 		allErrs = append(allErrs, validation.NewFieldRequired("volumeID"))
 	}
-	if cd.FSType == "" || (cd.FSType != "ext3" && cd.FSType != "ext4") {
-		allErrs = append(allErrs, validation.NewFieldRequired("fsType required and should be of type ext3 or ext4"))
+	if cd.FSType == "" {
+		allErrs = append(allErrs, validation.NewFieldRequired("fsType"))
```
Comment (swagiaal) [inline pkg/api/validation/validation.go]: I just noticed this. fsType is not required. mount.SafeFormatAndMount will default to ext4. In fact the parameter has "omitempty" in `types.go` even though the comment says required.

File: pkg/api/types.go [hunk 3]
```diff
@@ -575,7 +575,6 @@ type CinderVolumeSource struct {
 	VolumeID string `json:"volumeID"`
 	// Required: Filesystem type to mount.
```
Comment (swagiaal) [inline pkg/api/types.go]: Could you remove the comment above as fsType is not required.

**Comment by swagiaal** [comment]
Just a couple of nits but otherwise LGTM

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (jsafrane) [inline pkg/api/validation/validation.go]: I would do that in separate PR, it's 'omitempty' in almost all volume plugins, while the validator for all of them requires it's presence.

File: pkg/api/types.go (hunk 3 above)
Comment (jsafrane) [inline pkg/api/types.go]: Done.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (swagiaal) [inline pkg/api/validation/validation.go]: > I would do that in separate PR, it's 'omitempty' in almost all volume plugins, while the validator for all of them requires it's presence.

Ok.. sounds good

**Comment by jsafrane** [comment]
Another rebase.

**Comment by jsafrane** [comment]
@k8s-bot test this please

**Comment by swagiaal** [comment]
@k8s-bot unit test this

**Comment by ncdc** [comment]
@k8s-bot unit test this

**Comment by ncdc** [comment]
@jsafrane the build log indicates a compilation failure:

```
_output/local/go/src/k8s.io/kubernetes/pkg/volume/cinder/cinder_util_test.go:60: undefined: cinderSafeFormatAndMount
```

**Comment by jsafrane** [comment]
@k8s-bot unit test this

**Comment by jsafrane** [comment]
@k8s-bot test this

**Comment by swagiaal** [comment]
hmm.. I did not realize this still did not get merged :(
@jsafrane could you rebase please

**Comment by jsafrane** [comment]
Rebased.

**Comment by swagiaal** [comment]
@jsafrane thanks
@markturansky can you take a look at this ? it LGTM.

**Comment by markturansky** [comment]
This LGTM.  @jsafrane or @swagiaal, let me know when the local testing is done (I know Jan did it quite some time ago) and I'll add the label.

**Comment by jsafrane** [comment]
I tested it with today's master and there was actually a bug, `cinder_util.go` should use `FormatAndMount` instead of plain `Format`:

```diff
--- a/pkg/volume/cinder/cinder_util.go
+++ b/pkg/volume/cinder/cinder_util.go
@@ -80,7 +79,7 @@ func (util *CinderDiskUtil) AttachDisk(b *cinderVolumeBuilder, globalPDPath stri
                }
        }
        if notmnt {
-               err = b.blockDeviceMounter.Mount(devicePath, globalPDPath, b.fsType, options)
+               err = b.blockDeviceMounter.FormatAndMount(devicePath, globalPDPath, b.fsType, options)
                if err != nil {
                        os.Remove(globalPDPath)
                        return err

```

I pushed fixed patch and tested that Cinder volumes with `FsType="xfs"` are now automatically formatted with `xfs`, which was not possible before this patch.

**Comment by markturansky** [comment]
@jsafrane thanks, Jan.  Would that bug have caught by the unit test that was removed?  The test was originally for its own mounter, but could it be modified to use the mounter you're replacing it with?

**Comment by jsafrane** [comment]
The test tested the custom mounter I removed, not the plugin itself.

**Comment by markturansky** [comment]
Understood.  I think a unit test for the current mounter would be helpful, too.

**Comment by markturansky** [comment]
Thanks.  I'll have the confidence for LGTM with a unit test for the mounter.

**Comment by jsafrane** [comment]
OpenStack.AttachDisk is directly called from the cinder_util.go, so the test could run only with real OpenStack. No volume plugin is decoupled this way, unit testing of *_util.go files is not easily possible... We should make it better when rewriting the plugins for Attach controller.

**Comment by swagiaal** [comment]
Also, SafeFormatAndMount is fairly well unit tested independently of the plugins that use it.

**Comment by markturansky** [comment]
Ok.  SafeFormatAndMount certainly is unit tested and tested daily by everyone that mounts volumes.  Moving to it from a customer Cinder mounter is a good improvement.

**Comment by swagiaal** [comment]
@markturansky thanks for reviewing. @jsafrane good catch on that bug I can't believe I missed it having written the PR to rename Mount to FormatAndMount :/
